### ISSUE TITLE
[SDPA] cuDNN backward returns NaN grad_q for seq_len % 128 == 64 with attn_mask on H100/H200 (regression since 2.11, default backend)

### ISSUE THREAD

**Issue by qgallouedec** [pytorch/pytorch#196678]
### 🐛 Describe the bug

`scaled_dot_product_attention` with the cuDNN backend returns NaN in `grad_q` (forward, `grad_k`, `grad_v` are fine) when all three hold:

1. an explicit `attn_mask` is passed (boolean or float, an all-True mask triggers it too),
2. `seq_len % 128 == 64` (fails at 64, 192, 320, ..., 704, 832, 1088, 2112; clean at multiples of 128 and at every other remainder tested),
3. a query row's logsumexp is below about -88.7 = ln(FLT_MAX).

The padded half of the last 128-wide key tile is scored 0 instead of -inf -> `exp(0 - lse)` overflows fp32, -> inf * 0 (zero-padded K) gives NaN in dQ.

EFFICIENT_ATTENTION and MATH are clean on the same inputs, and so is cuDNN with `is_causal=True` and no mask tensor. Reproduces in bf16 and fp16, head_dim 32/64/128. The backward kernel that runs is `cudnn_generated_fort_native_sdpa_sm90_flash_bprop_wgmma_f16_knob_26_64x64x64_1x4x1`.

```python
import math, torch, torch.nn.functional as F
torch.manual_seed(0)
L, D = 704, 64  # fails for every L % 128 == 64; 640 and 768 are clean
q = torch.randn(1, 1, L, D, device="cuda", dtype=torch.bfloat16) * 0.1
k = torch.randn(1, 1, L, D, device="cuda", dtype=torch.bfloat16) * 0.1
v = torch.randn(1, 1, L, D, device="cuda", dtype=torch.bfloat16)
q[..., 0], k[..., 0] = 1.0, -120 * math.sqrt(D)  # every logit ~ -120, so every row's logsumexp < -88.7
q.requires_grad_(); k.requires_grad_(); v.requires_grad_()
mask = torch.ones(L, L, dtype=torch.bool, device="cuda").tril()[None, None]  # an all-True mask fails too
out = F.scaled_dot_product_attention(q, k, v, attn_mask=mask)
out.backward(torch.randn_like(out))
print("grad_q non-finite:", (~torch.isfinite(q.grad)).sum().item())
```

Version / hardware matrix (same script, PyPI wheels on HF Jobs except H100 which is our cluster with cu130 and cu129 wheels, L=704; ❌ = 45056 non-finite values in `grad_q`, ✅ = 0; L=640 and 768 are ✅ everywhere):
| torch (cuDNN) | T4 sm75 | A100 sm80 forced cuDNN | A10G sm86 forced cuDNN | L40S sm89 forced cuDNN | H100 sm90 default | H200 sm90 default | H200 sm90 forced cuDNN | RTX PRO 6000 sm120 forced cuDNN |
|---|---|---|---|---|---|---|---|---|
| 2.5.1 (9.1.0) | no cuDNN kernel | n/a (no cuDNN plan) | n/a (no cuDNN plan) | n/a (no cuDNN plan) | ✅ (efficient) | ✅ (efficient) | n/a (no cuDNN plan) | n/a (wheel lacks sm120) |
| 2.6.0 (9.1.0) | no cuDNN kernel | ❌ | ❌ | ❌ | ✅ (efficient) | ✅ (efficient) | ❌ | n/a (wheel lacks sm120) |
| 2.7.1 (9.5.1) | no cuDNN kernel | ❌ | ❌ | ❌ | ✅ (efficient) | ✅ (efficient) | ❌ | n/a (wheel lacks sm120) |
| 2.8.0 (9.10.2) | no cuDNN kernel | ❌ | ❌ | ❌ | ✅ (efficient) | ✅ (efficient) | ❌ | ❌ |
| 2.9.0 (9.10.2) | no cuDNN kernel | ❌ | ❌ | ❌ | ✅ (efficient) | ✅ (efficient) | ❌ | ❌ |
| 2.9.1 (9.10.2) | no cuDNN kernel | ✅ | ✅ | ✅ | ✅ (efficient) | ✅ (efficient) | ✅ | ✅ |
| 2.10.0 (9.10.2) | no cuDNN kernel | ✅ | ✅ | ✅ | ✅ (efficient) | ✅ (efficient) | ✅ | ✅ |
| 2.11.0 (9.19.0) | no cuDNN kernel | ✅ | ✅ | ✅ | ❌ (cudnn) | ❌ (cudnn) | ❌ | ✅ |
| 2.12.0 (9.20.0) | no cuDNN kernel | ✅ | ✅ | ✅ | ❌ (cudnn) | ❌ (cudnn) | ❌ | ✅ |
| 2.13.0 (9.20.0) | no cuDNN kernel | ✅ | ✅ | ✅ | ❌ (cudnn) | ❌ (cudnn) | ❌ | ✅ |

## Context

DPO/KTO fine-tuning of Qwen2.5-0.5B with HF transformers + TRL goes NaN after tens of steps on H100 with default settings (huggingface/trl#7181). Bisecting the failing micro-batch down to a single `scaled_dot_product_attention` call and replaying the captured q/k/v/mask through each backend isolated it to the cuDNN backward. The report below is the torch-only reduction.

### Versions

PyTorch version: 2.13.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCm SDK used to build PyTorch: N/A
HIP used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.13.13 | packaged by Anaconda, Inc. | (main, Apr 14 2026, 06:19:41) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-1057-aws-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.9.86
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA H100 80GB HBM3
Nvidia driver version: 580.159.04
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  96
On-line CPU(s) list:                     0-95
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 7R13 Processor
CPU family:                              25
Model:                                   1
Thread(s) per core:                      1
Core(s) per socket:                      48
Socket(s):                               2
Stepping:                                1
BogoMIPS:                                5299.99
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch topoext perfctr_core ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr rdpru wbnoinvd arat npt nrip_save vaes vpclmulqdq rdpid
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               3 MiB (96 instances)
L1i cache:                               3 MiB (96 instances)
L2 cache:                                48 MiB (96 instances)
L3 cache:                                384 MiB (12 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-47
NUMA node1 CPU(s):                       48-95
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; Safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Mitigation; Clear CPU buffers
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] nccl4py==0.4.1
[pip3] numpy==2.4.6
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cudnn-frontend==1.26.0
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.0.88
[pip3] nvidia-nvtx==13.0.85
[pip3] nvtx==0.2.15
[pip3] tokenspeed-triton==3.8.10.post20260721
[pip3] torch==2.13.0
[pip3] torch_c_dlpack_ext==0.1.5
[pip3] torchaudio==2.11.0
[pip3] torchcodec==0.15.0
[pip3] torchvision==0.28.0
[pip3] triton==3.7.1
[conda] nccl4py                                     0.4.1                pypi_0              pypi
[conda] numpy                                       2.4.6                pypi_0              pypi
[conda] nvidia-cublas                               13.1.1.3             pypi_0              pypi
[conda] nvidia-cuda-cupti                           13.0.85              pypi_0              pypi
[conda] nvidia-cuda-nvrtc                           13.0.88              pypi_0              pypi
[conda] nvidia-cuda-runtime                         13.0.96              pypi_0              pypi
[conda] nvidia-cudnn-cu13                           9.20.0.48            pypi_0              pypi
[conda] nvidia-cudnn-frontend                       1.26.0               pypi_0              pypi
[conda] nvidia-cufft                                12.0.0.61            pypi_0              pypi
[conda] nvidia-curand                               10.4.0.35            pypi_0              pypi
[conda] nvidia-cusolver                             12.0.4.66            pypi_0              pypi
[conda] nvidia-cusparse                             12.6.3.3             pypi_0              pypi
[conda] nvidia-cusparselt-cu13                      0.8.1                pypi_0              pypi
[conda] nvidia-nccl-cu13                            2.29.7               pypi_0              pypi
[conda] nvidia-nvjitlink                            13.0.88              pypi_0              pypi
[conda] nvidia-nvtx                                 13.0.85              pypi_0              pypi
[conda] nvtx                                        0.2.15               pypi_0              pypi
[conda] tokenspeed-triton                           3.8.10.post20260721  pypi_0              pypi
[conda] torch                                       2.13.0               pypi_0              pypi
[conda] torch-c-dlpack-ext                          0.1.5                pypi_0              pypi
[conda] torchaudio                                  2.11.0               pypi_0              pypi
[conda] torchcodec                                  0.15.0               pypi_0              pypi
[conda] torchvision                                 0.28.0               pypi_0              pypi
[conda] triton                                      3.7.1                pypi_0              pypi

cc @ezyang @gchanan @kadeng @msaroufim @csarofeen @ptrblck @eqy @nWEIdia @drisspg @liangel-02 @howardzhang-cv

**Comment by eqy** [comment]
We recently fixed a number of NaN issues in cuDNN, can you please try on latest nightlies?

**Comment by qgallouedec** [comment]
Thanks. Still present on latest nightly: `torch==2.15.0.dev20260909+cu130` (cuDNN 9.25.1, `torch.backends.cudnn.version()` = 92501), on both H100 and H200.

Same repro as above.
The output at L=704 is identical to 2.11 to 2.13 (same non-finite values, same rows), so whatever changed between 9.20 and 9.25 did not touch this path.
